### PR TITLE
initial attempt at supporting AES-GCM on PHP >= 7.1

### PR DESCRIPTION
**Pull request by fkooman**
Working on AES-GCM support for #134.

Tests are still missing. Requesting early feedback.

### CODE REVIEW

**Comment by vanderlee** [comment]
I'd like to have AES-GCM support included. Is there any way we can help or otherwise (financially) support the developers to include tested AES-GCM support?

**Comment by robrichards** [comment]
@vanderlee Are you able to generate any test documents that we can use to develop more tests from? I haven't had a chance to do this from other libraries yet so this would definitely help speed up adding support

**Comment by fkooman** [comment]
I have a Shibboleth encrypted aes256-gcm assertion 
[aes256.zip](https://github.com/robrichards/xmlseclibs/files/4369713/aes256.zip) that I use in [my SAML SP](https://github.com/fkooman/php-saml-sp/).

**Comment by vanderlee** [comment]
I'm trying to piece together a phpt testcase based on @fkooman's zip file but can't get it working. Is there any (online) tooling which should be used?

File: src/XMLSecurityKey.php [hunk 1]
```diff
@@ -397,12 +424,18 @@ private function unpadISO10126($data)
     private function encryptSymmetric($data)
     {
         $this->iv = openssl_random_pseudo_bytes(openssl_cipher_iv_length($this->cryptParams['cipher']));
-        $data = $this->padISO10126($data, $this->cryptParams['blocksize']);
-        $encrypted = openssl_encrypt($data, $this->cryptParams['cipher'], $this->key, OPENSSL_RAW_DATA | OPENSSL_ZERO_PADDING, $this->iv);
+        $authTag = null;
+        if(in_array($this->cryptParams, ['aes-128-gcm', 'aes-192-gcm', 'aes-256-gcm'])) {
```
Comment (robrichards) [inline src/XMLSecurityKey.php:428]: Should be:
if(in_array($this->cryptParams['cipher'], ['aes-128-gcm', 'aes-192-gcm', 'aes-256-gcm'])) {

File: src/XMLSecurityKey.php [hunk 2]
```diff
@@ -416,11 +449,19 @@ private function decryptSymmetric($data)
         $iv_length = openssl_cipher_iv_length($this->cryptParams['cipher']);
         $this->iv = substr($data, 0, $iv_length);
         $data = substr($data, $iv_length);
-        $decrypted = openssl_decrypt($data, $this->cryptParams['cipher'], $this->key, OPENSSL_RAW_DATA | OPENSSL_ZERO_PADDING, $this->iv);
+        $authTag = null;
+        if(in_array($this->cryptParams, ['aes-128-gcm', 'aes-192-gcm', 'aes-256-gcm'])) {
```
Comment (robrichards) [inline src/XMLSecurityKey.php:453]: Should be:
if(in_array($this->cryptParams['cipher'], ['aes-128-gcm', 'aes-192-gcm', 'aes-256-gcm'])) {

**Comment by robrichards** [comment]
@vanderlee Thanks for those test files. Am able to properly decrypt them now. Still need to test encryption and write some tests. Trying to find some time so might be another week or so.

**Comment by vanderlee** [comment]
@robrichards Tried to get it working with the changes you mentioned but not really getting anywhere with it. Please let me know if I can support this work in any way.

**Comment by robrichards** [comment]
@vanderlee Do you mind if I add your test file and key to the test suite? I did make independent encrypt/decrypt tests but having one created outside of the library would be very useful? I do have it fully working just looking at any other changes queued up for a release. Will be another day or 2.

**Comment by vanderlee** [comment]
@robrichards The testfiles aren't mine, but from @fkooman. @fkooman submitted this PR initially. All contents from the zipfile come straight from his own repo: https://github.com/fkooman/php-saml-sp/tree/master/tests/data, which has an MIT license so should be fine to use, though it's probably nice to have explicit permission from @fkooman.

**Comment by robrichards** [comment]
@fkooman Sorry about that. Forgot who had sent those :D. Anyways do you mind if I add that SAML file and the encryption key into this repo for a test case?

**Comment by fkooman** [comment]
As far as I'm concerned it is fine to add that assertion and (private) key, they are all test servers anyway!

**Comment by robrichards** [comment]
@fkooman I needed to fix a few things so used a diff from your branch and worked off of that. This has been released in 3.1.0

**Pull request closed without merging**